tsl/src/bgw_policy/policies_v2.c

```diff
@@ -0,0 +1,711 @@
+
+		refresh_policy tmp = {
+			.create_policy = true,
+			.start_offset = start_offset,
+			.end_offset = end_offset,
+			.schedule_interval = *DEFAULT_REFRESH_SCHEDULE_INTERVAL,
+			.start_offset_type = start_offset_type,
+			.end_offset_type = end_offset_type,
+		};
+		ref = tmp;
+		all_policies.refresh = &ref;
+	}
+	if (!PG_ARGISNULL(4))
+	{
+		compression_policy tmp = { .create_policy = true,
+								   .compress_after = PG_GETARG_DATUM(4),
+								   .compress_after_type = get_fn_expr_argtype(fcinfo->flinfo, 4) };
+		comp = tmp;
+		all_policies.compress = &comp;
+	}
+	if (!PG_ARGISNULL(5))
+	{
+		retention_policy tmp = { .create_policy = true,
+								 .drop_after = PG_GETARG_DATUM(5),
+								 .drop_after_type = get_fn_expr_argtype(fcinfo->flinfo, 5) };
+		ret = tmp;
+		all_policies.retention = &ret;
+	}
+	PG_RETURN_BOOL(validate_and_create_policies(all_policies, if_not_exists));
+}
+Datum
+policies_remove(PG_FUNCTION_ARGS)
+{
+	Oid cagg_oid = PG_GETARG_OID(0);
+	ArrayType *policy_array = PG_ARGISNULL(2) ? NULL : PG_GETARG_ARRAYTYPE_P(2);
+	bool if_exists = PG_GETARG_BOOL(1);
+	Datum *policy;
+	int npolicies, failures = 0;
+	int i;
+	bool success = false;
+
+	if (policy_array == NULL)
+		PG_RETURN_BOOL(false);
+
+	deconstruct_array(policy_array, TEXTOID, -1, false, TYPALIGN_INT, &policy, NULL, &npolicies);
+
+	for (i = 0; i < npolicies; i++)
+	{
+		char *curr_policy = VARDATA(policy[i]);
+
+		if (pg_strcasecmp(curr_policy, POLICY_REFRESH_CAGG_PROC_NAME) == 0)
+			success = policy_refresh_cagg_remove_internal(cagg_oid, if_exists);
+		else if (pg_strcasecmp(curr_policy, POLICY_COMPRESSION_PROC_NAME) == 0)
+			success = policy_compression_remove_internal(cagg_oid, if_exists);
+		else if (pg_strncasecmp(curr_policy,
+								POLICY_RETENTION_PROC_NAME,
+								strlen(POLICY_RETENTION_PROC_NAME)) == 0)
+			success = policy_retention_remove_internal(cagg_oid, if_exists);
+		else
+			ereport(NOTICE, (errmsg("No relevant policy found")));
+		if (!success)
+			++failures;
+	}
+	PG_RETURN_BOOL(success && (0 == failures));
+}
+
+Datum
+policies_remove_all(PG_FUNCTION_ARGS)
+{
+	if (PG_ARGISNULL(0))
+		PG_RETURN_BOOL(false);
+
+	Oid cagg_oid = PG_GETARG_OID(0);
+	bool if_exists = PG_GETARG_BOOL(1);
+	List *jobs;
+	ListCell *lc;
+	bool success = if_exists;
+	int failures = 0;
+	ContinuousAgg *cagg = ts_continuous_agg_find_by_relid(cagg_oid);
+
+	if (!cagg)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("\"%s\" is not a continuous aggregate", get_rel_name(cagg_oid))));
+
+	jobs = ts_bgw_job_find_by_hypertable_id(cagg->data.mat_hypertable_id);
+	foreach (lc, jobs)
+	{
+		BgwJob *job = lfirst(lc);
+		if (namestrcmp(&(job->fd.proc_name), POLICY_REFRESH_CAGG_PROC_NAME) == 0)
+			success = policy_refresh_cagg_remove_internal(cagg_oid, if_exists);
+		else if (namestrcmp(&(job->fd.proc_name), POLICY_COMPRESSION_PROC_NAME) == 0)
+			success = policy_compression_remove_internal(cagg_oid, if_exists);
```

Comment:
Do you have a test for this? It does not seems to be covered.

Reply:
Unfortunately I can't easily test this case.